src/Lib.P2P/Routing/DistributedQuery.cs

```diff
@@ -0,0 +1,327 @@
+        /// </summary>
+        /// <value>
+        ///   The default is 16.
+        /// </value>
+        /// <remarks>
+        ///   The number of peers that are asked for the answer.
+        /// </remarks>
+        public int ConcurrencyLevel { get; set; } = 16;
+
+        /// <summary>
+        ///   The distributed hash table.
+        /// </summary>
+        public DhtService Dht { get; set; }
+
+        /// <summary>
+        ///   The type of query to perform.
+        /// </summary>
+        internal MessageType QueryType { get; set; }
+
+        /// <summary>
+        ///   The key to find.
+        /// </summary>
+        internal MultiHash QueryKey { get; set; }
+
+        /// <summary>
+        ///   Starts the distributed query.
+        /// </summary>
+        /// <param name="cancel">
+        ///   Is used to stop the task.  When cancelled, the <see cref="TaskCanceledException"/> is raised.
+        /// </param>
+        /// <returns>
+        ///   A task that represents the asynchronous operation.
+        /// </returns>
+        public async Task RunAsync(CancellationToken cancel)
+        {
+            Log.Debug($"Q{Id} run {QueryType} {QueryKey}");
+
+            _runningQuery = CancellationTokenSource.CreateLinkedTokenSource(cancel);
+            Dht.Stopped += OnDhtStopped;
+            _queryMessage = new DhtMessage
+            {
+                Type = QueryType,
+                Key = QueryKey?.ToArray(),
+            };
+
+            var tasks = Enumerable
+               .Range(1, ConcurrencyLevel)
+               .Select(i =>
+                {
+                    var id = i;
+                    return AskAsync(id);
+                });
+            try
+            {
+                await Task.WhenAll(tasks).ConfigureAwait(false);
+            }
+            catch (Exception)
+            {
+                // eat it
+            }
+            finally
+            {
+                Dht.Stopped -= OnDhtStopped;
+            }
+
+            Log.Debug($"Q{Id} found {_answers.Count} answers, visited {_visited.Count} peers, failed {_failedConnects}");
+        }
+
+        private void OnDhtStopped(object sender, EventArgs e)
+        {
+            Log.Debug($"Q{Id} cancelled because DHT stopped.");
+            _runningQuery.Cancel();
+        }
+
+        /// <summary>
+        ///   Ask the next peer the question.
+        /// </summary>
+        private async Task AskAsync(int taskId)
+        {
+            var pass = 0;
+            var waits = 20;
+            while (!_runningQuery.IsCancellationRequested && waits > 0)
+            {
+                // Get the nearest peer that has not been visited.
+                var peer = Dht.RoutingTable
+                   .NearestPeers(QueryKey)
+                   .FirstOrDefault(p => !_visited.ContainsKey(p));
+                if (peer == null)
+                {
+                    --waits;
+                    await Task.Delay(100);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)